lib/Dialect/OM/Evaluator/Evaluator.cpp

```diff
@@ -0,0 +1,196 @@
+//===- Evaluator.cpp - Object Model dialect evaluator ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the Object Model dialect Evaluator.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/OM/Evaluator/Evaluator.h"
+#include "mlir/IR/BuiltinAttributeInterfaces.h"
+#include "mlir/IR/SymbolTable.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace mlir;
+using namespace circt::om;
+
+/// Construct an Evaluator with an IR module.
+circt::om::Evaluator::Evaluator(ModuleOp mod) : symbolTable(SymbolTable(mod)) {}
+
+/// Instantiate an Object with its class name and actual parameters.
+FailureOr<std::shared_ptr<Object>>
+circt::om::Evaluator::instantiate(StringAttr className,
+                                  ArrayRef<ObjectValue> actualParams) {
+  ClassOp cls = symbolTable.lookup<ClassOp>(className);
+  if (!cls)
+    return symbolTable.getOp()->emitError("unknown class name ") << className;
+
+  auto formalParamNames = cls.getFormalParamNames().getAsRange<StringAttr>();
+  auto formalParamTypes = cls.getBodyBlock()->getArgumentTypes();
+
+  // Verify the actual parameters are the right size and types for this class.
+  if (actualParams.size() != formalParamTypes.size()) {
+    auto error = cls.emitError("actual parameter list length (")
+                 << actualParams.size() << ") does not match formal "
+                 << "parameter list length (" << formalParamTypes.size() << ")";
+    error.attachNote() << "actual parameters: " << actualParams;
+    error.attachNote(cls.getLoc()) << "formal parameters: " << formalParamTypes;
+    return error;
+  }
+
+  // Verify the actual parameter types match.
+  for (auto [actualParam, formalParamName, formalParamType] :
+       llvm::zip(actualParams, formalParamNames, formalParamTypes)) {
+    Type actualParamType;
+    if (auto *attr = std::get_if<Attribute>(&actualParam))
+      if (auto typedActualParam = attr->dyn_cast_or_null<TypedAttr>())
+        actualParamType = typedActualParam.getType();
+    if (auto *object = std::get_if<std::shared_ptr<Object>>(&actualParam))
+      actualParamType = object->get()->getType();
+
+    if (!actualParamType || actualParamType != formalParamType) {
+      auto error = cls.emitError("actual parameter for ")
+                   << formalParamName << " has invalid type";
+      error.attachNote() << "actual parameter: " << actualParam;
```

Comment:
Does this crash when `actualPram` is nullptr?

Reply:
No, it will just cause actualParamType to never be set, and that will come through this error case. Although now you mention it, I think I will have a separate error message for the `!actualParamType` case explicitly, since this is a user facing API, and it's best to be clear.